controller/apps.go

```diff
@@ -0,0 +1,443 @@
+	// get the user definition (for cluster URL)
+	resp, err := authClient.ShowUser(ctx, authservice.ShowUserPath(), nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	status := resp.StatusCode
+	if status < 200 || status > 300 {
+		return nil, errors.New("Failed to GET user due to status code " + string(status))
+	}
+
+	respBody, err := ioutil.ReadAll(resp.Body)
+
+	var respType authservice.User
+	err = json.Unmarshal(respBody, &respType)
+	if err != nil {
+		return nil, err
+	}
+	return &respType, nil
+}
+
+func (c *AppsController) getToken(authClient authservice.Client, ctx context.Context, forService string) (*authservice.TokenData, error) {
+
+	resp, err := authClient.RetrieveToken(ctx, authservice.RetrieveTokenPath(), forService, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	status := resp.StatusCode
+	if status < 200 || status > 300 {
+		return nil, errors.New("Failed to GET user due to status code " + string(status))
+	}
+
+	respBody, err := ioutil.ReadAll(resp.Body)
+
+	var respType authservice.TokenData
+	err = json.Unmarshal(respBody, &respType)
+	if err != nil {
+		return nil, err
+	}
+	return &respType, nil
+}
+
+// getKubeClient createa kube client for the appropriate cluster assigned to the current user.
+// many different errors are possible, so controllers should call getAndCheckKubeClient() instead
+func (c *AppsController) getKubeClient(ctx context.Context) (*kubernetes.KubeClient, error) {
+
+	// create Auth API login object
+	authClient, err := auth.CreateClient(ctx, c.Config)
+	if err != nil {
+		log.Error(ctx, nil, "error accessing Auth server"+tostring(err))
+		return nil, err
+	}
+
+	authUser, err := c.getUser(*authClient, ctx)
+	if err != nil {
+		log.Error(ctx, nil, "error accessing Auth server"+tostring(err))
+		return nil, err
+	}
+
+	if authUser == nil || authUser.Data.Attributes.Cluster == nil {
+		log.Error(ctx, nil, "error getting user from Auth server:"+tostring(authUser))
+		return nil, errors.New("error getting user from Auth Server: %s" + tostring(authUser))
+	}
+
+	// get the login token for the cluster OpenShift API
+	osauth, err := c.getToken(*authClient, ctx, *authUser.Data.Attributes.Cluster)
+	if err != nil {
+		log.Error(ctx, nil, "error getting openshift credentials:"+tostring(err))
+		return nil, err
+	}
+
+	kubeURL := *authUser.Data.Attributes.Cluster
+	kubeToken := *osauth.AccessToken
+
+	kubeNamespaceName, err := c.getNamespaceName(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// create the cluster login object
+	kc, err := kubernetes.NewKubeClient(kubeURL, kubeToken, *kubeNamespaceName)
+	if err != nil {
+		return nil, err
+	}
+	return kc, nil
+}
+
+// getAndCheckKubeClient converts all errors Error 401, so errors can be returned from controllers as is
+func (c *AppsController) getAndCheckKubeClient(ctx context.Context) (*kubernetes.KubeClient, error) {
```

Comment:
why do you need this ? can't you return that from the `getKubeClient()` directly ?